modern-publish-plugin.md

```diff
@@ -0,0 +1,222 @@
+# Modern Gradle Publish Plugin
+
+
+## Goals
+
+- less magic configuration, more explicit APIs
+- provide APIs for everything represented through properties
+- remove or deprecate APIs that are already provided by Gradle
+- continue providing a way to get started with minimal configuration
+
+## New Structure
+
+For compatibility purposes the current `com.vanniktech.maven.publish` plugin
+behavior is kept as it is, with the exception of APIs that we decide to remove. 
+Most of the goals are solved through a set of new plugins for specific 
+platforms. These plugins provide APIs that replace the magic setup mechanisms
+of the current plugin. To avoid having to maintain two implentations the old
+plugin will internally just apply the new plugins and use their public APIs. 
+Over time we can decide if `com.vanniktech.maven.publish` should be deprecated
+or removed.
+
+### Plugins
+
+Specific plugins for each supported project type:
+- `com.vanniktech.maven.publish.java`
+- `com.vanniktech.maven.publish.gradle`
+- `com.vanniktech.maven.publish.android`
+- `com.vanniktech.maven.publish.kotlin.jvm`
+- `com.vanniktech.maven.publish.kotlin.js`
+- `com.vanniktech.maven.publish.kotlin.mpp`
```

Comment:
Another way to look at this would be to have one single plugin with `compileOnly` dependencies to all the aforementioned ones and let the user call appropriate methods later on:

```
mavenPublish {
   ...
   // creates a publication based on the "release" variant and empty javadoc/sources jars
   withAndroidPublications {
      // here there is a MavenPublication receiver to override behaviour if needed
      // e.g. if there is a sourcesJar available, one can do
      artifact(sourcesJar)
   }
} 
```

or in a kotlinMpp project:

```
mavenPublish {
   ...
   // does not create any publication as they are created by the kotlin plugin
   // Adds empty javadoc
   withKotlinMppPublications()
} 
```

Reply:
For a KMP library, only mpp plugin is necessary or if it has an android target, android and mpp are needed?

Reply:
@JavierMeep you'll need both `com.android.library` and `org.jetbrains.kotlin.multiplatform` in a MPP project using Android.

`org.jetbrains.kotlin.multiplatform` should create all publications though. IIRC, it misses javadoc (for all publications) and sources (for the metatdata one, maybe for android too)

Reply:
I know, but I am talking about these:

- `com.vanniktech.maven.publish.android`
- `com.vanniktech.maven.publish.kotlin.mpp`

I think a kmp library with Android target should use only one plugin vanniktech plugin.